pyvrp/cpp/SubPopulation.cpp

```diff
@@ -0,0 +1,156 @@
+{
+    for (auto &item : items)
+        delete item.individual;
+}
+
+void SubPopulation::add(Individual const *individual)
+{
+    // Copy the given individual into a new memory location, and use that from
+    // now on.
+    individual = new Individual(*individual);
+    Item item = {&params, individual, 0.0, {}};
+
+    for (auto &other : items)  // update distance to other individuals
+    {
+        auto const div = divOp(data, *individual, *other.individual);
+        auto cmp = [](auto &elem, auto &value) { return elem.first < value; };
+
+        auto &oProx = other.proximity;
+        auto place = std::lower_bound(oProx.begin(), oProx.end(), div, cmp);
+        oProx.emplace(place, div, individual);
+
+        auto &iProx = item.proximity;
+        place = std::lower_bound(iProx.begin(), iProx.end(), div, cmp);
+        iProx.emplace(place, div, other.individual);
+    }
+
+    items.push_back(item);
+    updateFitness();
+
+    if (size() > params.maxPopSize())
+        purge();
+}
+
+size_t SubPopulation::size() const { return items.size(); }
+
+SubPopulation::Item const &SubPopulation::operator[](size_t idx) const
+{
+    return items[idx];
+}
+
+std::vector<SubPopulation::Item>::const_iterator SubPopulation::cbegin() const
+{
+    return items.cbegin();
+}
+
+std::vector<SubPopulation::Item>::const_iterator SubPopulation::cend() const
+{
+    return items.cend();
+}
+
+void SubPopulation::remove(
+    std::vector<SubPopulation::Item>::iterator const &iterator)
+{
+    for (auto &[params, individual, fitness, proximity] : items)
+        // Remove individual from other proximities.
+        for (size_t idx = 0; idx != proximity.size(); ++idx)
+            if (proximity[idx].second == iterator->individual)
+            {
+                proximity.erase(proximity.begin() + idx);
+                break;
+            }
+
+    delete iterator->individual;  // dispose of manually allocated memory
+    items.erase(iterator);        // before the item is removed.
+}
+
+void SubPopulation::purge()
+{
+    while (size() > params.minPopSize)
+    {
+        // Remove duplicates from the subpopulation (if they exist)
+        auto const pred = [&](auto &iterator)
+        {
+            return !iterator.proximity.empty()
+                   && *iterator.proximity[0].second == *iterator.individual;
+        };
+
+        auto const duplicate = std::find_if(items.begin(), items.end(), pred);
+
+        if (duplicate == items.end())  // there are no more duplicates
+            break;
+
+        remove(duplicate);
+    }
+
+    while (size() > params.minPopSize)
+    {
+        // Remove the worst individual (worst in terms of biased fitness)
+        updateFitness();
+
+        auto const worstFitness = std::max_element(
+            items.begin(),
+            items.end(),
+            [](auto const &a, auto const &b) { return a.fitness < b.fitness; });
+
+        remove(worstFitness);
+    }
+}
+
+void SubPopulation::updateFitness()
```

Comment:
Maybe add some comment on when this is expected to be called. I think it's 'eager' for now rather than 'lazy' in the sense that it should always be up to date after a change, rather than before a use of the fitness.

If indeed the policy is 'eager', shouldn't we trigger this function directly once penalties change? Now it relies on an individual being added, which means we probably do a binary tournament first with cost based on old penalties. Now, I think that makes negligible difference in practice, but it is not what I'd expect and rather subtle.

Reply:
>Maybe add some comment on when this is expected to be called. I think it's 'eager' for now rather than 'lazy' in the sense that it should always be up to date after a change, rather than before a use of the fitness.

That's a good idea, I'll do this in a sec.

>If indeed the policy is 'eager', shouldn't we trigger this function directly once penalties change?

I think so, but... that updating is entirely internal to the `PenaltyManager` right now. It'd be quite a bit of work to make it possible to listen to that event, and then update our fitness vector based on it. Since this has been an issue going back to HGS-CVRP (I checked just yet, to be sure), I'm inclined to leave this as "wontfix".

I'll add a TODO at the top of the biasedFitness function about this.